File: loggalitic/src/main/java/com/android/oleksandrpriadko/loggalitic/logger/DefaultLogger.java
package com.android.oleksandrpriadko.loggalitic.logger;

import android.util.Log;

import com.android.oleksandrpriadko.loggalitic.Loggalitic;
import com.android.oleksandrpriadko.loggalitic.analytics.policy.Policy;

/**
 * wifi_switcher
 * Created by Oleksandr Priadko.
 * 7/9/17
 */

public class DefaultLogger extends Logger {

    public DefaultLogger(Policy policy) {
        super(policy);
    }

    @Override
    public boolean isLoggable(String tag, int level, boolean force) {
        return getPolicy().isLogAllowed(tag, level, force);
    }

    @Override
    public boolean isLogReportable(String tag, int level, boolean force) {
        return getPolicy().isLogReportAllowed(tag, level, force);
    }

    @Override
    public void v(String tag, String message, boolean force) {
        if (isLoggable(tag, Log.VERBOSE, force)) {
            Log.v(tag, message);
        }
        if (isLogReportable(tag, Log.VERBOSE, force)) {
            Loggalitic.publisher().event(tag, message);
        }
    }

    @Override
    public void d(String tag, String message, boolean force) {
        if (isLoggable(tag, Log.DEBUG, force)) {
            Log.d(tag, message);
        }
        if (isLogReportable(tag, Log.DEBUG, force)) {
            Loggalitic.publisher().event(tag, message);
        }
    }

    @Override
    public void i(String tag, String message, boolean force) {
        if (isLoggable(tag, Log.INFO, force)) {
            Log.i(tag, message);
        }
        if (isLogReportable(tag, Log.INFO, force)) {
            Loggalitic.publisher().event(tag, message);
        }
    }

    @Override
    public void w(String tag, String message, boolean force) {
        if (isLoggable(tag, Log.WARN, force)) {
            Log.w(tag, message);
        }
        if (isLogReportable(tag, Log.WARN, force)) {
            Loggalitic.publisher().event(tag, message);
        }
    }

    @Override
    public void e(String tag, String message, boolean force) {
        if (isLoggable(tag, Log.ERROR, force)) {
            Log.e(tag, message);
        }
        if (isLogReportable(tag, Log.ERROR, force)) {
            Loggalitic.publisher().event(tag, message);
        }
    }

    @Override
    public void v(String tag, String message) {
        v(tag, message, false);
    }

    @Override
    public void d(String tag, String message) {
        d(tag, message, false);
    }

    @Override
    public void i(String tag, String message) {
        i(tag, message, false);
    }

    @Override
    public void w(String tag, String message) {
        w(tag, message, false);
    }

    @Override
    public void e(String tag, String message) {
        e(tag, message, false);
    }

    @Override
    public void wtf(String tag, String message) {
        Log.wtf(tag, message);
    }
}

File: loggalitic/src/main/java/com/android/oleksandrpriadko/loggalitic/logger/DefaultLogger.kt
package com.android.oleksandrpriadko.loggalitic.logger

import android.util.Log
import com.android.oleksandrpriadko.core.logger.Logger
import com.android.oleksandrpriadko.core.policy.Policy
import com.android.oleksandrpriadko.loggalitic.LogPublishService

class DefaultLogger(policy: Policy) : Logger(policy) {

    override fun isLoggable(tag: String?, level: Int, force: Boolean): Boolean {
        return if (tag.isNullOrEmpty()) {
            false
        } else {
            policy.isLogAllowed(tag, level, force)
        }
    }

    override fun isLogReportable(tag: String?, level: Int, force: Boolean): Boolean {
        return if (tag.isNullOrEmpty()) {
            false
        } else {
            return policy.isLogReportAllowed(tag, level, force)
        }
    }

    override fun v(tag: String?, message: String?, force: Boolean) {
        requestReportAndSend(tag, message, force, Policy.VERBOSE)
    }

    override fun d(tag: String?, message: String?, force: Boolean) {
        requestReportAndSend(tag, message, force, Policy.DEBUG)
    }

    override fun i(tag: String?, message: String?, force: Boolean) {
        requestReportAndSend(tag, message, force, Policy.INFO)
    }

    override fun w(tag: String?, message: String?, force: Boolean) {
        requestReportAndSend(tag, message, force, Policy.WARN)
    }

    override fun e(tag: String?, message: String?, force: Boolean) {
        requestReportAndSend(tag, message, force, Policy.ERROR)
    }

    override fun wtf(tag: String?, message: String?, force: Boolean) {
        requestReportAndSend(tag, message, force, Policy.ASSERT)
    }

    private fun requestReportAndSend(tag: String?,
                                     message: String?,
                                     force: Boolean,
                                     @Policy.LogLevel logLevel: Int) {
        if (isLoggable(tag, logLevel, force)) {
            when (logLevel) {
                Policy.VERBOSE -> Log.v(tag, message)
                Policy.DEBUG -> Log.d(tag, message)
                Policy.INFO -> Log.i(tag, message)
                Policy.WARN -> Log.w(tag, message)
                Policy.ERROR -> Log.e(tag, message)
                Policy.ASSERT -> Log.wtf(tag, message)
                else -> {
                }
            }
        }
        if (isLogReportable(tag, logLevel, force)) {
            if (tag != null && message != null) {
                LogPublishService.publisher().event(tag, message)
            }
        }
    }

    override fun v(tag: String?, message: String?) {
        v(tag, message, false)
    }

    override fun d(tag: String?, message: String?) {
        d(tag, message, false)
    }

    override fun i(tag: String?, message: String?) {
        i(tag, message, false)
    }

    override fun w(tag: String?, message: String?) {
        w(tag, message, false)
    }

    override fun e(tag: String?, message: String?) {
        e(tag, message, false)
    }

    override fun wtf(tag: String?, message: String?) {
        Log.wtf(tag, message)
    }
}
